Download thread

Def0lti opened the pull request:
Добавлена функция скачки текста треда

Commit:
Добавлена функция скачки текста треда

Commit:
Изменен json файл, добавлен PathToDownloadDirectory

Commit:
Изменено время записанное в названии папке и внутри треда на UTC+4

VikSing88 commented on MattermostBot/appsettings.json:
```diff
@@ -3,6 +3,7 @@
 	"AccessToken": "xoxb-2344320041938-2368062522496-DXPQtHykKzWH3iWni9YAF4Fg",
 	"BotUserID": "U02AU1UFCEL",
 	"ChannelCheckPeriodInMinutes": "60",
+	"PathToDownloadDirectory": "D:\\TestDownload",
```
Давай сделаем один прямой слеш вместо двух обратных. Винда их нормально прожевывает.

VikSing88 commented on MattermostBot/appsettings.json:
```diff
@@ -3,6 +3,7 @@
 	"AccessToken": "xoxb-2344320041938-2368062522496-DXPQtHykKzWH3iWni9YAF4Fg",
 	"BotUserID": "U02AU1UFCEL",
 	"ChannelCheckPeriodInMinutes": "60",
+	"PathToDownloadDirectory": "D:\\TestDownload",
```
Давай сделаем один прямой слеш вместо двух обратных. Винда их нормально прожевывает.

Def0lti replied:
Исправлено
Но мне не нравится, как выглядит сообщение о скачке треда, если использовать прямой слеш в конфиге.
![Untitled1](https://user-images.githubusercontent.com/98971740/166882330-884e7483-770b-4465-a40b-aa22deb18efc.png)

Какая то путаница получается, первый слеш в пути прямой, а остальные обратные (Это Path.Combine обратные вставляет, тут мои полномочия всё)

+если в проводнике винды посмотреть путь до папки, то он тоже отображается с обратными слешами 
![Untitled](https://user-images.githubusercontent.com/98971740/166881950-290fc148-47e3-4e4b-992c-08ad9ca3e9e9.png)

А если использовать 2 обратных, то все красивенько получается (Путь выводится с одним обратным, их 2 для адекватного считывания строки)
![Untitled3](https://user-images.githubusercontent.com/98971740/166882855-354937d3-1190-48fa-9f85-9f8b909e6d3c.png)

Def0lti replied:
Если что, все работает нормально и так и так, если просто менять путь в конфиге, больше ничего трогать не нужно
Главное не использовать один обратный

VikSing88 replied:
Можно добавить Path.GetFullPath(Path.Combine(...)). 
См. https://stackoverflow.com/questions/3144492/how-do-i-get-nets-path-combine-to-convert-forward-slashes-to-backslashes

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    }
+
+    ///<summary>
+    ///Скачать тред
```
Общие замечания по оформлению:
1. Предпочтительнее объявлять переменные через var, а не указанием типа переменной (если в этом нет необходимости) - так проще читается код. Например:
`var path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.ToLocalTime());`
2. Поправить комментарии к методам:
- не везде есть описаны параметры методов;
- между символами "///" и дальнейшим текстом должен быть пробел, после описания метода должна ставиться точка. См. инструкцию по оформлению или код выше.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    }
+
+    ///<summary>
+    ///Скачать тред
```
Общие замечания по оформлению:
1. Предпочтительнее объявлять переменные через var, а не указанием типа переменной (если в этом нет необходимости) - так проще читается код. Например:
var path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.ToLocalTime());
2. Поправить комментарии к методам:
- не везде есть описаны параметры методов;
- между символами "///" и дальнейшим текстом должен быть пробел, после описания метода должна ставиться точка. См. инструкцию по оформлению или код выше.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///</summary>
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
+    {
+      string rootDateTimeString = Convert.ToString(rootDateTime).Replace(':', '-');
```
Есть вариант проще и понятнее:
`var rootDateTimeString = rootDateTime.ToString("yyyy.MM.dd hh-mm");`

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///</summary>
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
+    {
+      string rootDateTimeString = Convert.ToString(rootDateTime).Replace(':', '-');
```
Есть вариант проще и понятнее:
`var rootDateTimeString = rootDateTime.ToString("yyyy.MM.dd hh-mm");`

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
+    {
+      string rootDateTimeString = Convert.ToString(rootDateTime).Replace(':', '-');
+      string path = pathToDownloadDirectory + @"\" + string.Format("{0} {1} {2} {3}", rootDateTimeString, rootUserInfo.firstName, rootUserInfo.lastName, rootUserInfo.userName);
```
Есть удобная функция для формирования пути:
`Path.Combine(pathToDownloadDirectory, string.Format(...))`

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
+    {
+      string rootDateTimeString = Convert.ToString(rootDateTime).Replace(':', '-');
+      string path = pathToDownloadDirectory + @"\" + string.Format("{0} {1} {2} {3}", rootDateTimeString, rootUserInfo.firstName, rootUserInfo.lastName, rootUserInfo.userName);
```
Есть удобная функция для формирования пути:
`Path.Combine(pathToDownloadDirectory, string.Format(...))`

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///<summary>
+    ///Создать путь для скачки треда.
+    ///</summary>
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
```
CreateThreadPath звучит так, будто создается директория треда. Лучше бы метод назвать как-то так: GetPathToSaveThread.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///<summary>
+    ///Создать путь для скачки треда.
+    ///</summary>
+    private static string CreateThreadPath(string pathToDownloadDirectory, UserInfo rootUserInfo, DateTime rootDateTime)
```
CreateThreadPath звучит так, будто создается директория треда. Лучше бы метод назвать как-то так: GetPathToSaveThread.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+
+
+    ///<summary>
+    ///Создать путь для скачки треда.
```
Скачка - это про другое) Давай лучше будем использовать слово "скачивание".

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+
+
+    ///<summary>
+    ///Создать путь для скачки треда.
```
Скачка - это про другое) Давай лучше будем использовать слово "скачивание".

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    {
+      Message[] messages = mattermostApi.GetThreadMessages(messageEventInfo.rootID);
+      IEnumerable<Message> messagesSorted = GetMessagesSorted(RemoveDupesFromMessages(messages));
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
```
В messageEventInfo есть ИД пользователя, лучше его брать оттуда.

Def0lti replied:
Исправлено

Def0lti replied:
Вернул как было, если так сделать, в названии папки указывается имя того, кто сохраняет тред, а не того, кто его начал.

VikSing88 replied:
Ага, действительно, косяк)

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    {
+      Message[] messages = mattermostApi.GetThreadMessages(messageEventInfo.rootID);
+      IEnumerable<Message> messagesSorted = GetMessagesSorted(RemoveDupesFromMessages(messages));
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
```
В messageEventInfo есть ИД пользователя, лучше его брать оттуда.

VikSing88 commented on MattermostBot/IApiClient.cs:
```diff
@@ -54,6 +64,27 @@ public class MessageEventInfo
+    /// <summary>
+    /// Дата и время создания сообщения, от которого образован тред.
+    /// </summary>
+    public string rootDateTime;
```
Это поле добавлено, но нигде не означивается и не используется. 
Оно может пригодиться в методе DownloadThread (в комментарии к этому методу укажу, где rootDateTime можно заиспользовать), поэтому предлагаю это поле означить при создании нового MessageEventInfo.

Def0lti replied:
Удалил это поле. Объяснение в комментарии к DownloadThread

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -172,8 +187,7 @@ private Message[] GetMessagesByRequest(string request)
       postList.List = postList.Convert(j).List;
-      return postList.List.Select(p =>
-        new Message { messageId = p.id, dateTime = p.create_at ?? DateTime.MinValue, userId = p.user_id }).ToArray();
+      return postList.List.Select(p => new Message { messageId = p.id, dateTime = p.create_at ?? DateTime.MinValue, userId = p.user_id, message = p.message, fileIDs = p.file_ids }).ToArray();
```
Слишком длинная строка получается, надо перенести ее часть на следующую строку.

Def0lti replied:
Исправлено

VikSing88 replied:
Перенос получился не слишком удачным, т.к. оказалось разбитым выражение "dateTime = p.create_at ?? DateTime.MinValue". Было бы лучше оставить его целиком на одной строке.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -172,8 +187,7 @@ private Message[] GetMessagesByRequest(string request)
       postList.List = postList.Convert(j).List;
-      return postList.List.Select(p =>
-        new Message { messageId = p.id, dateTime = p.create_at ?? DateTime.MinValue, userId = p.user_id }).ToArray();
+      return postList.List.Select(p => new Message { messageId = p.id, dateTime = p.create_at ?? DateTime.MinValue, userId = p.user_id, message = p.message, fileIDs = p.file_ids }).ToArray();
```
Слишком длинная строка получается, надо перенести ее часть на следующую строку.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      IEnumerable<Message> messagesSorted = GetMessagesSorted(RemoveDupesFromMessages(messages));
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
+
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
```
1) Вместо messages[0].dateTime лучше заиспользовать messageEventInfo.rootDateTime. 
2) Вместо AddHours(4) лучше использовать ToLocalTime().

Def0lti replied:
Исправлено на messages[0].dateTime.ToLocalTime().
Т.к. rootDateTime должен быть полем messageEventInfo, его нужно добавить как таковое в MattermostApiAdapter, однако что бы это сделать, надо в MattermostApiAdapter вызвать список сообщений, что бы из него взять корневое. В DownloadThread это делается через mattermostApi.GetThreadMessages, а в MattermostApiAdapter нельзя использовать mattermostApi (клиент для работы). Следовательно и корневое сообщение получить нельзя, ну либо нужно для этого делать какой то отдельный метод, а это, как мне кажется, того не стоит.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      IEnumerable<Message> messagesSorted = GetMessagesSorted(RemoveDupesFromMessages(messages));
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
+
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
```
1) Вместо messages[0].dateTime лучше заиспользовать messageEventInfo.rootDateTime. 
2) Вместо AddHours(4) лучше использовать ToLocalTime().

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
+      string pathFilesFolder = path + @"\files";
+      Directory.CreateDirectory(pathFilesFolder);
+      string threadTxt = path + @"\thread.txt";
```
1. Тут лучше заиспользовать Path.Combine().
2. Имя переменной  сбивает с толку. Надо назвать в соответствии с тем, что туда будет записано. Например, pathToFile.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
+      string pathFilesFolder = path + @"\files";
+      Directory.CreateDirectory(pathFilesFolder);
+      string threadTxt = path + @"\thread.txt";
```
1. Тут лучше заиспользовать Path.Combine().
2. Имя переменной  сбивает с толку. Надо назвать в соответствии с тем, что туда будет записано. Например, pathToFile.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      foreach (Message message in messagesSorted)
+      {
+        var userInfo = mattermostApi.GetUserInfoByID(message.userId);
+        thread.WriteLine(message.dateTime.AddHours(4) + "\n" + userInfo.firstName + " " + userInfo.lastName + " (" + userInfo.userName + "):"+ message.message + "\n");
```
1. AddHours(4) надо заменить на ToLocalTime().
2. Вывод полного имени и никнейма в txt (и в имени папки кстати тоже) кажется лишним. В Directum Talk у нас автоматически заполнено полное имя (FirstName + LastName), можно его выводить. А если вдруг его нет (на всякий случай), то только тогда выводить никнейм.

Def0lti replied:
Исправлено

VikSing88 replied:
Логика по определению имени/никнейма пользователя используется в 2 местах: здесь и в методе GetPathToSaveThread. Хочется для этой логики выделить отдельный метод. Тогда при необходимости внести изменения, нам понадобится подправить код только в одном месте.
И еще несколько моментов: 
1) Скобки для никнейма точно нужны?
2) После двоеточий при выводе сообщений в файл нужны пробелы.
3) Некорректно отрабатывает условие if (userInfo.firstName != null && userInfo.lastName != null), если у пользователя не заполнены firstName и lastName.

Def0lti replied:
1) Сейчас уже нет, забыл их убрать. Исправлено
2) Исправлено
3) Исправлено, заменил на !string.IsNullOrEmpty()

Def0lti replied:
Сделал отдельные методы для получения имени, фамилии и никнейма

VikSing88 replied:
> Сделал отдельные методы для получения имени, фамилии и никнейма

Я имела в виду, что надо выделить один метод, в котором будет анализироваться, есть ли у пользователя имя/никнейм. В результате этот метод должен выдавать строку с именем/никнеймом, которая потом будет использоваться при выводе текста треда в файл и в методе GetPathToSaveThread.
Когда добавишь этот метод, уйдет много лишнего кода. 
Методы GetUserFirstName и т.д. тоже не нужны.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      foreach (Message message in messagesSorted)
+      {
+        var userInfo = mattermostApi.GetUserInfoByID(message.userId);
+        thread.WriteLine(message.dateTime.AddHours(4) + "\n" + userInfo.firstName + " " + userInfo.lastName + " (" + userInfo.userName + "):"+ message.message + "\n");
```
1. AddHours(4) надо заменить на ToLocalTime().
2. Вывод полного имени и никнейма в txt (и в имени папки кстати тоже) кажется лишним. В Directum Talk у нас автоматически заполнено полное имя (FirstName + LastName), можно его выводить. А если вдруг его нет (на всякий случай), то только тогда выводить никнейм.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+        thread.WriteLine(message.dateTime.AddHours(4) + "\n" + userInfo.firstName + " " + userInfo.lastName + " (" + userInfo.userName + "):"+ message.message + "\n");
+      }
+      thread.Close();
+      Console.WriteLine("Тред скачан по адресу: " + threadTxt);
```
Сообщение о том, что тред скачан, надо писать не в консоль (там его никто не увидит), а выводить пользователю "эфемерным" сообщением или сообщением в тред.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+        thread.WriteLine(message.dateTime.AddHours(4) + "\n" + userInfo.firstName + " " + userInfo.lastName + " (" + userInfo.userName + "):"+ message.message + "\n");
+      }
+      thread.Close();
+      Console.WriteLine("Тред скачан по адресу: " + threadTxt);
```
1. Сообщение о том, что тред скачан, надо писать не в консоль (там его никто не увидит), а выводить 
пользователю "эфемерным" сообщением или сообщением в тред.
2. Предлагаю заменить формулировку выводимого сообщения на: "Тред скачан и находится здесь: <..>". При этом путь надо указывать не до txt-файла, а до папки. У нас ведь там еще будут файлы лежать.

Def0lti replied:
Исправлено

VikSing88 replied:
Кажется, что вывод в консоль уже лишнее, достаточно эфемерного сообщения пользователю.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
+
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
+      string pathFilesFolder = path + @"\files";
```
Имхо, создавать папку для файлов не стоит заранее: если файлов в треде не будет, то смысла в ней нет.
П.С. В слакботе сейчас папка создается всегда, но, думаю, это поведение не стоит один в один переносить сюда.

Def0lti replied:
Пока просто убрал эту папку в принципе
Тут надо сперва со скачкой файлов разобраться

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      UserInfo rootUserInfo = mattermostApi.GetUserInfoByID(messages[0].userId);
+
+      string path = CreateThreadPath(pathToDownloadDirectory, rootUserInfo, messages[0].dateTime.AddHours(4));
+      string pathFilesFolder = path + @"\files";
```
Имхо, создавать папку для файлов не стоит заранее: если файлов в треде не будет, то смысла в ней нет.
П.С. В слакботе сейчас папка создается всегда, но, думаю, это поведение не стоит один в один переносить сюда.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
           break;
         }
       }
+      else 
```
Если пользователь изначально написал неправильно команду и потом ее отредактировал, то такая команда не сработает, Роберто ее проигнорирует. Тут дело в методе StartNewPostEventHandling: там есть только обработка новых сообщений ("posted").
Есть два варианта, что с этим делать:
1. В методе StartNewPostEventHandling получать не только новые сообщения, но и отредактированные. 
2. Анализировать сообщения в треде в этапа: сначала смотреть, есть ли там упоминание имени бота, а потом анализировать текст команды. Если команда написана с ошибкой (например, @roberto dowload), то выводить сообщение пользователю, что бот такой команды не знает, а знает только команду "download".
Второй вариант кажется предпочтительнее - во-первых, указывает пользователю на ошибку (вдруг он не заметил), во-вторых, можно будет добавлять другие команды помимо download.

Def0lti replied:
Тут проблема в том, что метод Contains учитывает регистр при сравнении, и ему нельзя сказать, что бы он так не делал, в отличие от Regex.
Так что лучше сперва инициализировать наличие символа @ используя Contains, а после этого проверять строку через Regex на факт обращения к Роберто и вызова существующей команды.

Def0lti replied:
Исправлено

VikSing88 replied:
>  метод Contains учитывает регистр при сравнении, и ему нельзя сказать, что бы он так не делал

Можно) См. https://docs.microsoft.com/ru-ru/dotnet/api/system.string.contains?view=net-6.0.

`Contains(str, StringComparison.CurrentCultureIgnoreCase)`

VikSing88 replied:
Проблема текущего кода в том, что он пропускает сообщение вида: "@ , roberto download".

VikSing88 replied:
Давай зафиксируем требования к формату команды (надо было это сделать раньше, ну как есть). 
Формат: @<имя_бота> <команда>.
Примерный формат команд брала из документации к hubot: [здесь](https://github.com/ketchalegend/hubot-Mattermost) и [здесь ](https://github.com/loafoe/hubot-matteruser).
1) Не принимаем в расчет регистр команды. 
2) Если в тексте команды прописаны некоторые знаки препинания, то они игнорируются (например, можно игнорировать : и ,). Пример: `@roberto, download`
3) Игнорируем лишние пробелы, в т.ч. в начале текста.
4) Команда может состоять из нескольких слов, разделенных пробелом. Это задел на будущее. Пример: `@roberto autopin off`

VikSing88 replied:
Теперь по реализации. 
1) (это я уже писала, наверное ты не увидел). Имя бота лучше получать через его ИД, указанном в конфиге: `mattermostApi.GetUserInfoByID(botUserID).userName`
Зашивать его имя нельзя, т.к. в Directum Talk оно немного отличается от нашего тестового.
2) Первоначальный фильтр должен быть все-таки по имени бота (т.е. "@<имя_бота>"), иначе работает не так, как нам надо. Contains/Regex - тут на твое усмотрение. 
3) Пока разбиралась, накидала тестовый проект, как выуживать команду из сообщения к боту. Вдруг тебе пригодится: G:\Lebedeva_VN\Roberto Regex.

Def0lti replied:
Исправлено.
Внедрил твой проект.

VikSing88 replied:
1. Строку "@roberto" нельзя зашивать в коде. В одном месте ты уже заиспользовал mattermostApi.GetUserInfoByID(botUserID).userName, в остальных надо тоже самое сделать (вернее один раз сохранить значение в переменную и ее использовать везде).
2. Пока тестировала, наткнулась вот на такое сообщение от бота:
![image](https://user-images.githubusercontent.com/58815407/169973073-2afffb0b-be0a-4dda-9286-a0e96c4382bc.png)
Давай, чтоб таких коллапсов не было, из текста сообщения от бота уберем введенную пользователем команду. И слово "Ошибка", оно тут лишнее. Тут кстати можно сделать сообщение от имени бота ("Я не знаю такой команды. Доступные команды: ...") или можно оставить так, как и было, немного только поменяв текстовку ("Роберто не знает такой команды. Доступные команды: ...").

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
           break;
         }
       }
+      else 
```
Если пользователь изначально написал неправильно команду и потом ее отредактировал, то такая команда не сработает, Роберто ее проигнорирует. Тут дело в методе StartNewPostEventHandling: там есть только обработка новых сообщений ("posted").
Есть два варианта, что с этим делать:
1. В методе StartNewPostEventHandling получать не только новые сообщения, но и отредактированные. 
2. Анализировать сообщения в треде в этапа: сначала смотреть, есть ли там упоминание имени бота, а потом анализировать текст команды. Если команда написана с ошибкой (например, @roberto dowload), то выводить сообщение пользователю, что бот такой команды не знает, а знает только команду "download".
Второй вариант кажется предпочтительнее - во-первых, указывает пользователю на ошибку (вдруг он не заметил), во-вторых, можно будет добавлять другие команды помимо download.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+        }
+        catch (DirectoryNotFoundException dirEx)
+        {
+          Console.WriteLine("Путь не найден: " + dirEx.Message);
```
Обработка исключения должна быть не здесь, а в методе, где создается эта папка. И надо будет выводить пользователю, что во время скачивания треда произошла ошибка. А то он будет не в курсе, что пошло не так.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+        }
+        catch (DirectoryNotFoundException dirEx)
+        {
+          Console.WriteLine("Путь не найден: " + dirEx.Message);
```
Обработка исключения должна быть не здесь, а в методе, где создается эта папка. И надо будет выводить пользователю, что во время скачивания треда произошла ошибка. А то он будет не в курсе, что что-то пошло не так.

Def0lti replied:
Исправлено

VikSing88 replied:
Пользователю надо выводить сообщение общего вида (например, "Произошла ошибка при скачивании треда. Обратитесь к администратору."), полный стек ошибки пользователю ни к чему. А вот в консоль уже можно выводить подробности.
И еще: имхо лучше не использовать кучу восклицательных знаков и пугающих эмодзи в сообщениях пользователю - вдруг он от ужаса происходящего за сердце схватится)))

VikSing88 replied:
П.С. При возникновении DirectoryNotFoundException бот выводит НЕэфемерное сообщение в канал и сам же его пинит)

VikSing88 replied:
![Pasted image 20220516122234](https://user-images.githubusercontent.com/58815407/168768947-3d0230e8-8330-4301-8884-88611d0e1ff7.png)

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      {
+        try 
+        {
+          if (messageEventInfo.message.Contains('@') == true)
```
1. Лучше наверное проверять, что сообщение начинается с @<имя бота>. Текущая проверка "пропускает" на дальнейший анализ все сообщения, где хоть кто-то упоминается через @.
2. Имя бота лучше получать через его ИД, указанном в конфиге:  `mattermostApi.GetUserInfoByID(botUserID).userName`
Зашивать его имя нельзя, т.к. в Directum Talk оно немного отличается от нашего тестового.

VikSing88 replied:
п.2 не сделан. 
Продублировала этот пункт в другом комментарии, а этот закрою.
UPD: github не дает закрыть этот комментарий пока ты здесь не отпишешься. Пока оставляю.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///</summary>
+    private static IEnumerable<Message> RemoveDupesFromMessages(Message[] messages)
+    {
+      var messagesNoDupes = messages.GroupBy(x => x.messageId).Select(y => y.First());
```
Хочется перенести логику формирования корректного списка сообщений, отсортированного в нужном порядке, в класс MattermostApiAdapter.
И надо будет там прописать комментарий, что маттермост выдает "дубликат" первого сообщения, поэтому нам надо его удалить.

Def0lti replied:
Оставил как есть, по той же причине, что удалил rootDateTime.
Добавил комментарий.

VikSing88 replied:
> Оставил как есть, по той же причине, что удалил rootDateTime.

Перенос фильтрации сообщений никак не связан с rootDateTime.
Я предлагала объединить методы GetMessagesSorted и RemoveDupesFromMessages и этот объединенный метод перенести в модуль MattermostApiAdapter. Так лучше сделать потому, что выдача корректного списка сообщений - это ответственность класса MattermostApiAdapter.

Def0lti replied:
Перенес функционал RemoveDupesFromMessages в GetMessagesSorted, перенес в MattermostApiAdapter.
Исправлено.

VikSing88 replied:
1. Давай вызов GetMessagesSorted сделаем из метода GetThreadMessages, т.к. GetThreadMessages нам уже должен выдать корректный список сообщений. Тогда GetMessagesSorted можно сделать приватным и убрать модификатор static. 
2. Доп. комментарий " .GroupBy необходим т.к. Mattermost присылает первое сообщение дважды, нужно от него избавиться." надо записывать по такому формату:
```
    /// <summary>
    /// Отсортировать тред.
    /// </summary>
    /// <param name="messages">Список сообщений в треде.</param>
    /// <remarks"> GroupBy необходим т.к. Mattermost присылает первое сообщение дважды, нужно от него избавиться.</remarks>
```
И еще: сортировка - это расположение элементов в определенном порядке. Тут скорее фильтрация. Надо поправить коммент.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,6 +226,74 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    ///</summary>
+    private static IEnumerable<Message> GetMessagesSorted(IEnumerable<Message> messages)
+    {
+      IEnumerable<Message> messagesSorted = messages.OrderBy(message => message.dateTime);
```
В этом методе и методе ниже локальная переменная кажется лишней. Можно проще написать:
`return messages.OrderBy(message => message.dateTime);`

Def0lti replied:
Исправлено

VikSing88 reviewed (changes requested):
На доработку. Если что-то непонятно. можем обсудить.

Commit:
Исправления после Review

Commit:
При скачке треда теперь скачиваются и  файлы

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -62,6 +70,14 @@ public void PostMessage(string channelID, string message, string rootId = null)
       Post.Create(api, channelID, message, rootId);
     }
 
+    public async void CreateLinkedFile(string messageID, string file_id, string pathToFile)
```
По логике. 
Оказалось, что имена скачиваемых файлов могут быть не уникальными, и тогда при попытке создать файл с таким же именем, валится ошибка. Придется для файлов формировать уникальные имена. Есть 2 варианта: 1) проверять, если файл уже есть, то добавить постфикс в имя (как это делает винда), например: имя_файла(1).jpg; 2) в качестве имени брать ID и приплюсовывать оригинальное расширение. Если учесть, что некоторые картинки вставляются с именем "Изображение вставлено...", то второй вариант кажется предпочтительнее. Или может у тебя будут варианты, как лучше это сделать.

По оформлению. 
1. Лучше вызывать метод скачивания файлов DownloadLinkedFiles здесь:
	
```
          if (message.fileNames != null)
          {
            foreach (var name in message.fileNames)
            {
              thread.WriteLine("Добавлен файл - {0}\n", name.ToString());
            }
          }
```

Тогда метод DownloadThread станет более компактным.
2. Кажется, что вытаскивать имена файлов из ММ проще тут же (констуркция if из п.1). Я имею в виду, что из метода MattermostApiAdapter.GetMessagesByRequest надо убрать вызов метода GetLinkedToPostFileNames, т.к. нам нужны имена файлов только в определенном случае - при скачивании треда.

Def0lti replied:
Логику направил, сделал названия как ты предложила по второму варианту. 
Но, второй пункт оставил как есть. Методу GetLinkedToPostFileNames нужен сам Post, который выдается из ММ и существует только в MattermostApiAdapter.

Def0lti replied:
Можно конечно сделать исключение для одного метода, но мне это как то странно кажется, учитывая, что у нас в Program никаких данных напрямую из ММ не идет, все проходит через адаптер. Надо обсудить наверное.

Def0lti replied:
Исправил

VikSing88 replied:
Обсудили. 
Есть замечания по реализации CreateLinkedFile. Написала в другом комментарии. Этот закрываю.

VikSing88 reviewed (changes requested):
На доработку.

Commit:
Изменения после ревью

Commit:
Убрал GetLinkedToPostFileNames из MattermostApiAdapter.GetMessagesByRequest.

Убрал поле fileNames у message
Имена файлов теперь вытаскиются толко там, где они нужны в Program.cs

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,12 +231,151 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+            thread.WriteLine(message.dateTime.ToLocalTime() + "\n" + GetUserFirstName(userInfo) + " " + GetUserLastName(userInfo) + ": " + message.message + "\n");
+          else
+            thread.WriteLine(message.dateTime.ToLocalTime() + "\n" + GetUserUsername(userInfo) + ": " + message.message + "\n");
+          if(message.fileIDs != null)
```
Здесь предлагаю сделать как-то так:
```
        if (message.fileIDs != null)
            message.fileIDs
              .ToList()
              .ForEach(fileId =>
              {
                 var fileName = mattermostApi.GetFileById(message.messageId, fileId, pathToFilesFolder).Result;
                thread.WriteLine("Добавлен файл - {0}\n", fileName);
              });
```
Метод GetFileByIdэто практически CreateLinkedFile, только плюсом там надо сначала получить имя файла через GetFileInfo. Примерно так:
```
   public async Task<string> GetFileById(string messageID, string fileID, string pathToFile)
    {
      var post = Post.GetById(api, messageID).Result;
      var originalFileName = post.GetFileInfo(api, fileID).Result.name;
      var newFileName = fileID + Path.GetExtension(originalFileName);
      var file = File.Create(Path.Combine(pathToFile, fileID + Path.GetExtension(newFileName)));
      await post.GetFile(api, fileID, file);
      file.Close();
      return newFileName;
    }
```

Def0lti replied:
Исправлено

VikSing88 reviewed (changes requested):
На доработку.

Commit:
Исправления после ревью + добавил сообщение о начале скачивания треда

VikSing88 commented on MattermostBot/IApiClient.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Threading;
 using System.Threading.Tasks;
+using MattermostApi;
```
Лишнее.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/IApiClient.cs:
```diff
@@ -69,6 +103,11 @@ public interface IApiClient
     public void PostMessage(string channelID, string message, string rootID = null);
 
+    ///<summary>
+    ///Получить данные пользователя по ID.
```
Пробел.
Плюс надо дописать в комментариях, что возвращает метод (в блоке "returns").

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -62,9 +71,31 @@ public void PostMessage(string channelID, string message, string rootId = null)
+
+
+    /// <summary>
+    /// Отфильтровать тред.
```
1. Лучше написать комментарий в такой формулировке: "Отфильтровать лишние сообщения в треде".
2. Нужен блок "returns".
3. Метод  нужно переименовать, а то имя не соответствует комментарию. Можно что-то типа GetFilteredMessages.
4. Приватный метод надо перенести к остальным приватным методам. Из МИ-02.14:
"Внутри региона следует сортировать члены в порядке видимости. Сперва – публичные члены, затем – более закрытые".

Def0lti replied:
Исправлено

VikSing88 replied:
Блок returns должен идти перед remarks.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -92,10 +123,17 @@ public void PostEphemeralMessage(string channelID, string userID, string message
     async public void StartReceivingServerMessages(Action<MessageEventInfo> newPostEventHandler, Action<string> errorEventHandler, CancellationToken cancellationToken)
     {
+      try
+      {
```
Отступы внутри блока.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -95,6 +102,11 @@ private class MessageInfo
     private static int channelCheckPeriodInMinutes;
 
+    ///<summary>
+    ///Путь к папке с сохраненными тредами
```
Отступ.

Def0lti replied:
Исправлено

VikSing88 replied:
Раз уж есть и другие мелочи и все равно надо будет делать коммит, то давай здесь точку еще поставим)

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,12 +231,108 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+          }
+          else
+            mattermostApi.PostEphemeralMessage(messageEventInfo.channelID, messageEventInfo.userID, string.Format(
+              "Я не знаю такой команды.\nДоступные команды: {0}", "download"));
```
Вместо строки "download" лучше использовать константу downloadCommand.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,12 +231,108 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+          thread.WriteLine(message.dateTime.ToLocalTime() + "\n" + GetUserName(userInfo) + ": " + message.message + "\n");
+          if (message.fileIDs != null)
+          {
+            if (filesFolderExist == false)
```
Давай вместо того, чтобы каждый раз в цикле проверять, что у нас директория уже создана, мы сделаем перед циклом foreach так:
```
if (messages.Where(m => m.fileIDs != null).Count() > 0)
          Directory.CreateDirectory(pathToFilesFolder);
```

Def0lti replied:
Исправлено

VikSing88 replied:
Надо немного привести в порядок очередность действий:
```
        var pathToFilesFolder = Path.Combine(pathToThread, "files");  <- здесь объявляем переменную
        Directory.CreateDirectory(pathToThread);
        var pathToThreadFile = Path.Combine(pathToThread, "thread.txt");
        StreamWriter thread = File.CreateText(pathToThreadFile); 
        if (messages.Where(m => m.fileIDs != null).Count() > 0)
          Directory.CreateDirectory(pathToFilesFolder);  <- только здесь ее используем
```

Давай примерно такой порядок сделаем:
1. Создаем директорию для треда.
2. Создаем директорию для файлов, если необходимо.
3. Создаем файл для треда.

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,12 +231,108 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+      }
+      else
+      {
+        return userInfo.userName.ToString();
```
ToString() - лишнее. 
userName уже строка.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/Program.cs:
```diff
@@ -218,12 +231,108 @@ private static void NewPostEventHandler(MessageEventInfo messageEventInfo)
+    /// <returns>Строка с полным именем пользователя. Если нет имени или фамилии, возвращаетя никнейм.</returns>
+    private static string GetUserName(UserInfo userInfo)
+    {
+      if (!string.IsNullOrEmpty(userInfo.firstName) && !string.IsNullOrEmpty(userInfo.lastName))
```
Давай в метод передавать не userInfo, а ИД пользователя. В обоих случаях, когда используется метод GetUserName, мы сначала получаем userInfo и затем передаем его в метод GetUserName. Лишний код получается.

Def0lti replied:
Исправлено

VikSing88 commented on MattermostBot/MattermostApiAdapter.cs:
```diff
@@ -62,9 +71,31 @@ public void PostMessage(string channelID, string message, string rootId = null)
     public Message[] GetThreadMessages(string postID)
     {
-      return GetMessagesByRequest(Api.Combine("posts", postID, "thread"));
+      return  GetMessagesSorted(GetMessagesByRequest(Api.Combine("posts", postID, "thread"))).ToArray();
```
Лишний пробел после слова return.

Def0lti replied:
Исправлено

VikSing88 reviewed (changes requested):
Мелкие замечания по оформлению. 
На доработку.

Commit:
Исправления по оформлению

VikSing88 reviewed (changes requested):
Еще немного.

Commit:
Еще немного исправлений по оформлению

Commit:
Merge remote-tracking branch 'origin/master' into download-thread

# Conflicts:
#	MattermostBot/IApiClient.cs
#	MattermostBot/MattermostApiAdapter.cs
#	MattermostBot/Program.cs
#	MattermostBot/appsettings.json
